client/bench_test.go

```diff
@@ -0,0 +1,210 @@
+package bench
+
+import (
+	matmult "github.com/amwolff/2-lg-mats-bench/gen/go/amwolff/matmult/v1"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/grpc"
+	"io/ioutil"
+	"log"
+	"math"
+	"os"
+	"testing"
+)
+
+type initVariables struct {
+	client   matmult.MatrixProductAPIClient
+	request  matmult.MultiplyRequest
+	response matmult.MultiplyResponse
+}
+
+var initVars initVariables
+
+func TestMain(m *testing.M) {
+	conn, err := grpc.Dial(serverAddr, grpc.WithInsecure(), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(math.MaxInt64)))
+	if err != nil {
+		log.Fatalf("Couldn't connect to the server: %v\n", err)
+	}
+	client := matmult.NewMatrixProductAPIClient(conn)
+
+	pbFile, err := os.Open(pbPath)
+	if err != nil {
+		log.Fatalf("Couldn't open protool buffers message: %v\n", err)
+	}
+
+	pbBytes, err := ioutil.ReadAll(pbFile)
+	if err != nil {
+		log.Fatalf("Couldn't read protool buffers message: %v\n", err)
+	}
+
+	var request matmult.MultiplyRequest
+	if err := proto.Unmarshal(pbBytes, &request); err != nil {
+		log.Fatalf("Couldn't get request message: %v\n", err)
+	}
+
+	response, err := multiplyOnSiteCPP(client, &request)
+	if err != nil {
+		log.Fatalf("Couldn't get a response: %v\n", err)
+	}
+
+	initVars.client = client
+	initVars.request = request
+	initVars.response = *response
+
+	exit := m.Run()
+	conn.Close()
+	pbFile.Close()
+	os.Exit(exit)
+}
+
+func BenchmarkPbInPbOut(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		_, err := multiplyOnSiteCPP(initVars.client, &initVars.request)
```

Comment:
Because the benchmark doesn't affect the global state in any way (although I'm not sure about the call to `b.Errorf`), the compiler might optimise your code as such:

```
func BenchmarkPbInPbOut(b *testing.B) {
    for i := 0; i < b.N; i++ {

        }
}
```

[Watch out for compiler optimisations](https://dave.cheney.net/high-performance-go-workshop/dotgo-paris.html#watch_out_for_compiler_optimisations).

BTW this whole workshop is really, really good.